### ISSUE TITLE
ansible-galaxy

### ISSUE THREAD

**Issue by itwars** [k3s-io/k3s-ansible#39]
Hi all,
Here is my proposal: 

I will probably create a new branch to test the ansible-galaxy directory structure and reorganize the structure!

Are you ok with that way to do it?

**Comment by paulfantom** [comment]
If you need any pointers, I was initially refactoring those roles with aim to put them in ansible-galaxy at https://github.com/rancher/k3s/pull/1649, but don't have time right now to get back to that.

Those refactored roles are currently used by me in https://github.com/thaum-xyz/ankhmorpork/tree/master/ansible

**Comment by itwars** [comment]
Ok I will dive into your code, then create a new branch, and after testing this branch will be perhaps merge to master :)

**Comment by emansom** [comment]
@ebesson @jpraczyk @xanmanning @onedr0p are all working on similiar goals. What would be the proper way to form a consensus?

**Comment by itwars** [comment]
Let's talk! What about a slack? Just send me your email to my email and I will be please to make it!

vincent.rabah@gmail.com

**Comment by xanmanning** [comment]
Emailed 😊

**Comment by ebesson** [comment]
Emailed blush

**Comment by forty** [comment]
Any progress on this?

**Comment by rdxmb** [comment]
Seems that the master changed too quickly and the changes for galaxy too slow - because IMHO this seems a re-structuring of that repository.

I could help doing that, but parallel changes in the master will get merge conflicts very quickly.

**Comment by rdxmb** [comment]
My suggestion is that we take some time to refactor that - while the commits to master should have a break.

@itwars what do you think?

**Comment by itwars** [comment]
Yes I agree, for this purpose I've create the branch ansible-galaxy

**Comment by rdxmb** [comment]
Also, I would like to welcome @geerlingguy to help doing that, for he is a very experienced ansible-guy :+1:

**Comment by onedr0p** [comment]
Hi everyone, I would like to point out that there is [another k3s galaxy role](https://github.com/PyratLabs/ansible-role-k3s) that is already in ansible-galaxy and works wonderfully. It would be nice if these could merge into one.

**Comment by rdxmb** [comment]
Sounds good, I will have a look at it.

In a comment above also https://github.com/rancher/k3s-ansible/issues/39#issuecomment-639331969 https://github.com/thaum-xyz/ankhmorpork/tree/master/ansible is mentioned.

**Comment by onedr0p** [comment]
@rdxmb my repo uses @xanmanning k3s role under the hood. It's an opinionated way to provision Ubuntu, k3s, CNIs and other things I wanted.

**Comment by xanmanning** [comment]
> Hi everyone, I would like to point out that there is [another k3s galaxy role](https://github.com/PyratLabs/ansible-role-k3s) that is well maintained, already in ansible-galaxy and works wonderfully. It would be nice if these could merge into one here. Maybe the maintainer there, @xanmanning could comment.

Thanks for your kind words @onedr0p. I have been asked a number of times about working in this role rather than my own or even creating PRs to merge my code in.

Unfortunately I'm quite constrained by time at the moment with a lot going on personally and some health issues as well. I did join the Slack group kindly set up my @itwars however time constraints have made it difficult to put in any valuable contribution on my part.

My role was created as a way of me advancing my career, mostly to write better quality Ansible but to also learn about Kubernetes. I'm still extremely new to Kubernetes as I don't work with it daily. This is why I display the warning about use of my role in production.

There are also valid reasons I can see for people not wanting to pick my role given the direction I'm intending on taking. I picked k3s as a project because at the time I don't think anyone else had created a role. I started a day or two after general release so as to not get influenced by other people's existing code so I can build the quality of my own code. I have always welcomed PRs though as this helps me learn and I've had some fantastic contributions. As this is a method for me to train myself I'm actually looking to go into the direction of implementing more validation checks into my role, which is intended to tell the operator of the role that everything is configured as expected and any deviations can be identified. The next step after that is to start looking at security best practices, implementing and validating those in the role. To me these are a sign of quality, however this may not be everyone's "cup of tea" (pardon the British expression).

**TL;DR** I'm under a lot of time constraints and I'm still in the early phases of my Ansible and Kubernetes journey so I'm not confident I can deliver real value.

**Comment by rdxmb** [comment]
Thanks, @xanmanning 

As this is in the rancher-namespace, would somebody from rancher coordinate the refactoring-process, as I wrote above with pausing/blocking the commits to master or something like that?

@itwars As you created this project, could you do that?

**Comment by rdxmb** [comment]
seems there is no interest in doing that. Actually, I do not have time and priority for that anymore.

**Comment by ayr-ton** [comment]
Are you folks open for related PRs?

**Comment by xanmanning** [comment]
Anyone fancy taking over as maintainer of my Ansible role? I don't have a lot of time to maintain it any more :disappointed: .

https://galaxy.ansible.com/xanmanning/k3s (Currently ranked number 1 k3s role by downloads [on Ansible Galaxy](https://galaxy.ansible.com/search?keywords=k3s&order_by=-relevance&deprecated=false&type=role&page=1)).

Sorry to hijack, but if there's future interest in putting an official k3s Ansible role into galaxy I've got a fair following already. :shrug:

**Comment by jon-stumpf** [comment]
@xanmanning , I will look into your project and let you know.

**Comment by ayr-ton** [comment]
@xanmanning I can take it. @jon-stumpf Do you want to pair on this?

**Comment by ayr-ton** [comment]
If make sense we could also merge anything into here.

**Comment by onedr0p** [comment]
It would be a shame to see @xanmanning work be unmaintained, I vote we migrate his work here if any maintainer here is up for it.

**Comment by ayr-ton** [comment]
@xanmanning How would you like us to proceed?

**Comment by xanmanning** [comment]
I'm probably not going to be much help in terms of direction as I'm really just looking for an out and wanted to offer the opportunity for the code to be maintained by a bigger community, be it the k3s team or the k8s@home folk. Otherwise I'm just going to have to abandon the project 

Happy for any features you like to be picked out into this repo, or for ownership of my repo to be transferred and features from this repo merged in. 

Probably depends on interest, workloads and appetite as to which route is taken and I'm keen to try and assist with the transition.

**Comment by ayr-ton** [comment]
@xanmanning I volunteer. ✋ 

I can start the process by both reviewing the existing PRs in your repo and selecting which features make sense to start PRing here until k3s-io folks are confident with the changes or with taking it as the official repo. Until there, I will be happy to keep everything updated in your namespace. o/

Feel free to contact me via ayrton at riseup.net or via GitHub issues/discussions.

**Comment by onedr0p** [comment]
@xanmanning if all else fails you could make @ayr-ton and I co-maintainers of your current repo, that's if you don't mind us bugging you once a year if something comes up ;)

**Comment by ayr-ton** [comment]
@onedr0p count on me to help with the PRs revision effort o/
So, reviewing the current plan:
- Give a burst on the waiting revisions so we could move forward with current fixes and implementations
- Once we get a stable flow, start making PRs from there to here.

On my side, I will give preference to anything that improves the flow for contributors (CI for the playbook, etc) or existing playbook flows that don't break any contract or idempotency.

**Comment by onedr0p** [comment]
> Once we get a stable flow, start making PRs from there to here.

Honestly I feel we cannot make that decision until we get a maintainer of this on-board, without that our efforts might be for nothing.

I strongly feel the best course of action would be to have a @k3s-io team member fork @xanmanning ansible repo into a new repository and deprecate this one.

**Comment by jon-stumpf** [comment]
@onedr0p and @ayr-ton, I have reviewed all of the issues and PRs for this repo.  Many can be closed.  I have three PRs open and received no response to an email that I sent early last month.  It would be good to get a new repo open under **k3s-io** so we could start fresh.

**Comment by itwars** [comment]
Hi folks,

When I've initiate this issue, it was just for a matter of being compliant with ansible galaxy pattern: when you 
```
ansible-galaxy init dummy
```
It create a directories structure with empty yaml files: 

```
dummy      ────────────┬───────── vars
                       │
                       ├───────── tests
                       │
                       ├───────── templates
                       │
                       ├───────── tasks
                       │
                       ├───────── meta
                       │
                       ├───────── handlers
                       │
                       ├───────── files
                       │
                       └───────── defaults
```

So we must spread existing yaml files to thoses directories. And voilà!

**Comment by ayr-ton** [comment]
@itwars Seems that the discussion evolved. We could either keep discussing the ansible-galaxy scheme here or move the other discussions to other issues to keep this one focused.

Current threads:
- Ansible Galaxy support for the current repo (the original issue)
- https://galaxy.ansible.com/xanmanning/k3s maintainability as the most mature role we have for k3s instead of creating a new galaxy submission. If there are new topics/issues to discuss we can create an issue there.
- Integrate the code from `xanmanning/k3s` here in small PRs (starting from the galaxy feature) or maintain a fork here as part of `k3s-io` repos officially (could be a new issue, but the decision is up to `k3s-io/k3s-ansible` core committers) 

I'm comfortable with any directions. Is a matter of which opportunities both communities want to invest energy.

**Comment by jon-stumpf** [comment]
@onedr0p and @ayr-ton, do you want to discuss a roadmap?

**Comment by devinnasar** [comment]
Did this ever go anywhere? I see that https://github.com/PyratLabs/ansible-role-k3s still displays a help-wanted sign.
